starlite/middleware/csrf.py

```diff
@@ -0,0 +1,103 @@
+        csrf_cookie = request.cookies.get(self.config.cookie_name)
+        existing_csrf_token = request.headers.get(self.config.header_name)
+
+        send_f = send
+
+        if request.method not in self.config.safe_methods:
+            if not self._csrf_tokens_match(existing_csrf_token, csrf_cookie):
+                response = self._get_error_response()
+                await response(scope, receive, send)
+                return
+        else:
+
+            async def send_wrapper(message: Message) -> None:
+                if csrf_cookie is None and message["type"] == "http.response.start":
+                    message.setdefault("headers", [])
+                    headers = MutableHeaders(scope=message)
+                    if "set-cookie" not in headers:
+                        cookie: SimpleCookie = SimpleCookie()
+                        cookie[self.config.cookie_name] = self._generate_csrf_token()
+                        cookie[self.config.cookie_name]["path"] = self.config.cookie_path
+                        cookie[self.config.cookie_name]["secure"] = self.config.cookie_secure
+                        cookie[self.config.cookie_name]["httponly"] = self.config.cookie_httponly
+                        cookie[self.config.cookie_name]["samesite"] = self.config.cookie_samesite
+                        if self.config.cookie_domain is not None:
+                            cookie[self.config.cookie_name]["domain"] = self.config.cookie_domain
+                        headers.append("set-cookie", cookie.output(header="").strip())
+                await send(message)
+
+            send_f = send_wrapper
+
+        await self.app(scope, receive, send_f)
+
+    def _generate_csrf_hash(self, token: str) -> str:
+        return hmac.new(self.config.secret.encode(), token.encode(), hashlib.sha256).hexdigest()
+
+    def _generate_csrf_token(self) -> str:
+        token = secrets.token_hex(CSRF_SECRET_BYTES)
+        token_hash = self._generate_csrf_hash(token)
+        return token + token_hash
+
+    def _decode_csrf_token(self, token: str) -> str:
+        if len(token) < CSRF_SECRET_LENGTH + 1:
+            raise BadSignature
+
+        ts = token[:CSRF_SECRET_LENGTH]
+        existing_hash = token[CSRF_SECRET_LENGTH:]
+        expected_hash = self._generate_csrf_hash(ts)
+        if not secrets.compare_digest(existing_hash, expected_hash):
+            raise BadSignature
+
+        return ts
+
+    def _csrf_tokens_match(self, request_csrf_token: Optional[str], cookie_csrf_token: Optional[str]) -> bool:
+        if not (request_csrf_token and cookie_csrf_token):
+            return False
+        try:
+            decoded_request_token = self._decode_csrf_token(request_csrf_token)
+            decoded_cookie_token = self._decode_csrf_token(cookie_csrf_token)
+            return secrets.compare_digest(decoded_request_token, decoded_cookie_token)
+        except BadSignature:
+            return False
```

Comment:
So, please change thiis - you are using exceptions to do the if/else switching here. Rather, return None from the `_decode_csrf_token` method instead of raising an exception, and then return false if either `self._decode_csrf_token(request_csrf_token)` or `self._decode_csrf_token(cookie_csrf_token)` are `None`.

Reply:
Done!